impl/src/backend.test.ts

```diff
@@ -0,0 +1,152 @@
+import { fairlyAllocateCredit } from "./backend";
+
+import { strict as assert } from "assert";
+import test from "node:test";
+import * as simplestatistics from "simple-statistics";
+
+interface FairlyAllocateCreditTestCase {
+  name: string;
+  credit: number[];
+  value: number;
+  needsRand?: boolean;
+}
+
+function noRand(): number {
+  throw new Error("no rand expected");
+}
+
+type Interval = [min: number, max: number];
+
+// https://en.wikipedia.org/wiki/Normal_distribution#Quantile_function
+function normalPpf(q: number, mean: number, stdDev: number): number {
+  return (
+    mean +
+    stdDev * Math.sqrt(2) * simplestatistics.inverseErrorFunction(2 * q - 1)
+  );
+}
+
+function getIntervalApprox(n: number, p: number, alpha: number): Interval {
+  // Approximates a binomial distribution with a normal distribution which is a bit
+  // simpler as it is symmetric.
+  const mean = n * p;
+  const variance = mean * (1 - p);
+  const diff = normalPpf(1 - alpha / 2, 0, Math.sqrt(variance));
+  return [mean - diff, mean + diff];
+}
+
+function getAllIntervals(
+  n: number,
+  creditFractions: readonly number[],
+  alphaTotal: number,
+): Interval[] {
+  // When testing multiple hypotheses, dividing `alphaTotal` by the dimension
+  // for each individual test is sometimes called the "bonferroni correction".
+  const alpha = alphaTotal / creditFractions.length;
+  return creditFractions.map((cf) => getIntervalApprox(n, cf, alpha));
+}
+
+function runFairlyAllocateCreditTest(
+  tc: Readonly<FairlyAllocateCreditTestCase>,
+): void {
+  // TODO: replace with precise sum
+  const sumCredit = tc.credit.reduce((a, b) => a + b, 0);
+  const normalizedFloatCredit = tc.credit.map(
+    (item) => (tc.value * item) / sumCredit,
+  );
+
+  const [rand, k] = tc.needsRand ? [Math.random, 100000] : [noRand, 1];
```

Comment:
```suggestion
  const [rand, k] = tc.needsRand ? [Math.random, 1000] : [noRand, 1];
```

No point thrashing TOO hard if you have a solid statistical test.